### ISSUE TITLE
Parallel message processing

### ISSUE THREAD

**Issue by paolofulgoni** [chkr1011/MQTTnet#732]
### Describe the feature request

This is my use case:

* I'm using MQTT as a message bus between services
* A service subscribes to an event. To handle the event, the service must make an RPC to another service

This requires two message handler running in parallel, which is not allowed by the library because it causes a deadlock (see https://github.com/chkr1011/MQTTnet/issues/648#issuecomment-499997450)

### Which project is your feature request related to?

Client

### Describe the solution you'd like

I'd like the library to handle the messages in parallel as soon as they are received.

### Describe alternatives you've considered

Start a new `Task` in the message handler, to process the message in parallel.

### Example

I made a (failing) test which shows the deadlock issue:

https://github.com/paolofulgoni/MQTTnet/commit/ac27f2a8a849671673c3ec61d91b41965c5d3b33

**Comment by paolofulgoni** [comment]
This is related to #648, but not exactly the same. In #648 the issue was caused by the PUBACK required by QoS1.

**Comment by paolofulgoni** [comment]
I understand from the discussion in #648 that the parallel processing of the messages was avoided to honor the message ordering, but I don't clearly understand why.

Is it only because you want to send the PUBACK in the same order?

**Comment by chkr1011** [comment]
No the reason is the business logic of the hosting application. The first versions of MQTTnet had started a Task for every new incoming message. But the problem is that this leads to ordering issues and people start to complain that hardware for example behaves strange because the order is wrong (even small reordering were ciritical). So now it is one single thread to keep things ordered. But this also gives some benefit in memory consumption, responsiveness and reliability because a new message is only accepted if the previous one was accepted. Lets consider the old way. There will be plenty of threads started (or tasks queued) even if the system is already at 100 % load. There is no chance to recover and the server will send messages over and over. This does not happen when processing the messages in serial order (no matter which QoS is used).

Please also keep in mind that MQTT is NO RPC protocol. So sending another message is not an option. Please checkout the RPC extension I build. It uses different topics to identify request and reply. This allows having some kind of RPC but respecting the MQTT protocol specifications.

**Comment by paolofulgoni** [comment]
Thanks for the detailed explanation!

I understand the reasons which led to the current implementation and that the main use case is telemetry. Nevertheless there are certainly other use cases where a different behavior is preferred.

I wonder if the library could be configurable to allow these two independent options:

* Parallel message handling (like the old version you mentioned)
* send ack just before the message handler starts. That is, let the developer choose between ack on message received or ack on message handled

The first would have the side effect to brake the guarantee of message handling in reception order. The second one brakes that sort of flow control you mentioned, if I well understand.

With respect to RPC, I partly disagree, as it has been standardized in MQTTv5. Anyway the implementation would be similar to what you did in the RPC project, using a pair of correlated pub/sub messages.

Of course my use case, i.e. sending an RPC form a message handler, is probably quite uncommon. :-)

**Comment by PayDrum** [comment]
I'm currently facing a similar situation but not sure if this is the same. Here is my use-case:

I'm receiving telemetry data through MQTT, parsing the message payloads and inserting into database. Now since our queries are a bit expensive, I have a requirement to aggregate multiple datasets from messages and insert them all at the same time. Problem is that due to hard set requirement, I am not allowed to implement any sort of persistence; therefore, I need to only complete the messages(through processingfailed property), when the number of messages have been all successfully inserted and if any failed, I will set the processingfailed to true, so that the broker retries this later. 

My approach so far has been creating an Async countdownevent in my dataAggregator service, and making the insert when the countdownevent reaches 0, but for some reason that I cannot figure out for the life of me, after going through the first message, a deadlock is caused. There are no more messages passed on by the mqtt client even though I am awaiting a Task which should hand the control back and have the client pass on more messages.

Is this scenario supported by the current implementation?

**Comment by chkr1011** [comment]
MQTTv5 has for sure the correlation ID but this also affects a complete message. The problem in this case is that the QoS of a single message is used to handle RPC calls. 

I can add such settings but they are dangerous. When enabling parallel handling it may happen very quickly that your client becomes unresponsive. Starting a new Task is quite fast and receiving messages as well. If you receive 1,000 messages per second it will start 1,000 tasks per second and will never stop. Even if the system is already messed up. That was the reason to change over to serial processing. When adding some kind of limit to this we end up in the same dead lock.

**Comment by paolofulgoni** [comment]
I understand your point of view. But I think that the current implementation is a bit dangerous too, as it may cause a deadlock.

Anyway, I suppose that my use case, or similar situations where parallel message handling is needed, are not common. Therefore I'm closing this issue, let's see if anyone else will ever ask for this feature.

Thanks again for your support
